src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs

```diff
@@ -2193,7 +2193,7 @@ private int GatherClassGCLayout(TypeDesc type, byte* gcPtrs)
         {
             int result = 0;
 
-            if (type.IsByReferenceOfT)
+            if (type.IsByReferenceOfT || type.IsByRef)
```

Comment:
This wasn't failing the test but I knew this would need adjustment. There's equivalent code in the runtime JitInteface. I think there is a GC hole right now and we probably need a test:

https://github.com/dotnet/runtime/blob/5a9bcde1d52b8b5a6c7b70883704bcb349c54c12/src/coreclr/vm/jitinterface.cpp#L2033-L2082

Reply:
@MichalStrehovsky Can you give an example of a type that could create a GC hole?

Reply:
This is used by RyuJIT to compute GC information for locals.

```csharp
ref struct X
{
    ref int Y;
}

static void Method()
{
    X x = ...;
}
```

The generated GC tracking for `x` should be pretty much equivalent to the local being typed as `ref int`, but RyuJIT doesn't see into X and JitInterface is going to tell RyuJIT there's nothing GC needs to be aware of. I believe the generated stack reporting won't report the interor pointer within x.

Reply:
Right! Thank you for calling this out. I was updating [SOS yesterday](https://github.com/dotnet/diagnostics/pull/2829) and ran `gcinfo` on a few methods using types like above and I had assumed it would show the `ref` field. I couldn't figure out why it wasn't shown so have it on my list of items to investigate. Perhaps this is that issue. I'll take a look tomorrow.

Reply:
@MichalStrehovsky I don't think this is correct now that I look at it. The only reason I believe this branch exists is because we use a type (i.e., `ByReference<T>`) instead of metadata. The loop over fields below should handle this when a field is marked as `ref`. I think this branch will be removed once we delete the `ByReference<T>` type.

https://github.com/dotnet/runtime/blob/e312bda175d72267b0c5ada0a9c471313a2ae3da/src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs#L2221-L2224

Your point about CoreCLR stands though. The code in `jitinterface.cpp` does need to be updated. I am working on that.

Reply:
I've also update one of the tests and after enabling GCStress, it fails. Will investigate what needs to happen on the CoreCLR side.

Reply:
You're right, it's unnecessary. I spent more time looking at the VM voice than this code. We don't usually diverge but looks like this was already futureproofed here.